File: backend/app/services/matching.py
"""
매칭 서비스
가맹점명 → 사용내역 자동 매칭
"""
from typing import Optional, Tuple
from sqlalchemy.orm import Session

from app.repositories.pattern_repo import PatternRepository
from app.models.pattern import Pattern, MatchType


class MatchingService:
    """패턴 매칭 서비스"""

    def __init__(self, db: Session):
        self.db = db
        self.pattern_repo = PatternRepository(db)

    def find_match(
        self,
        merchant_name: str,
        card_id: Optional[int] = None,
    ) -> Tuple[Optional[str], Optional[int]]:
        """
        가맹점명으로 사용내역 찾기

        3단계 매칭 로직:
        1. 카드 전용 정확 매칭
        2. 공통 정확 매칭
        3. 포함(CONTAINS) 매칭

        Args:
            merchant_name: 가맹점명
            card_id: 카드 ID (카드 전용 패턴 우선 검색용)

        Returns:
            (사용내역, 패턴ID) 튜플. 매칭 없으면 (None, None)
        """
        pattern = self.pattern_repo.find_matching_pattern(merchant_name, card_id)

        if pattern:
            # 사용 횟수 증가
            self.pattern_repo.increment_use_count(pattern.id)
            return pattern.usage_description, pattern.id

        return None, None

    def create_pattern_from_manual(
        self,
        merchant_name: str,
        usage_description: str,
        card_id: Optional[int] = None,
        created_by: Optional[str] = None,
    ) -> Pattern:
        """
        수동 입력에서 새 패턴 생성

        Args:
            merchant_name: 가맹점명
            usage_description: 사용내역
            card_id: 카드 ID (None이면 공통 패턴)
            created_by: 생성자

        Returns:
            생성된 Pattern
        """
        return self.pattern_repo.create(
            merchant_name=merchant_name,
            usage_description=usage_description,
            card_id=card_id,
            match_type=MatchType.EXACT.value,
            priority=0 if card_id is None else 10,  # 카드 전용은 우선순위 높음
            created_by=created_by,
        )

    def get_match_stats(self) -> dict:
        """매칭 통계 조회"""
        all_patterns = self.pattern_repo.get_all()

        total = len(all_patterns)
        by_type = {}
        by_card = {}

        for p in all_patterns:
            # 타입별 분류
            match_type = p.match_type
            by_type[match_type] = by_type.get(match_type, 0) + 1

            # 카드별 분류
            card_key = "common" if p.card_id is None else f"card_{p.card_id}"
            by_card[card_key] = by_card.get(card_key, 0) + 1

        return {
            "total_patterns": total,
            "by_type": by_type,
            "by_card": by_card,
        }

    def batch_rematch(self, card_id: Optional[int] = None) -> dict:
        """
        미매칭 거래에 대해 재매칭 시도

        Args:
            card_id: 특정 카드만 재매칭 (None이면 전체)

        Returns:
            재매칭 통계
        """
        from app.models.transaction import Transaction, MatchStatus

        query = self.db.query(Transaction).filter(
            (Transaction.usage_description.is_(None)) |
            (Transaction.usage_description == "")
        )

        if card_id:
            query = query.filter(Transaction.card_id == card_id)

        pending = query.all()
        stats = {"total": len(pending), "matched": 0, "failed": 0}

        for tx in pending:
            usage, pattern_id = self.find_match(tx.merchant_name, tx.card_id)
            if usage:
                tx.usage_description = usage
                tx.matched_pattern_id = pattern_id
                tx.match_status = MatchStatus.AUTO.value
                stats["matched"] += 1
            else:
                stats["failed"] += 1

        self.db.commit()
        return stats

    def get_card_patterns(self, card_id: int) -> list:
        """카드별 패턴 목록 (공통 포함)"""
        from app.models.pattern import Pattern

        patterns = self.db.query(Pattern).filter(
            (Pattern.card_id == card_id) | (Pattern.card_id.is_(None))
        ).order_by(
            Pattern.card_id.desc(),  # 카드 전용 먼저
            Pattern.use_count.desc()
        ).all()

        return patterns

    def suggest_patterns(self, merchant_name: str, card_id: Optional[int] = None) -> list:
        """
        가맹점명으로 유사 패턴 제안

        Args:
            merchant_name: 가맹점명
            card_id: 카드 ID

        Returns:
            유사 패턴 목록
        """
        from app.models.pattern import Pattern

        # 가맹점명 일부가 포함된 패턴 검색
        query = self.db.query(Pattern)
        if card_id:
            query = query.filter(
                (Pattern.card_id == card_id) | (Pattern.card_id.is_(None))
            )

        all_patterns = query.all()
        suggestions = []

        for p in all_patterns:
            # 부분 매칭 점수 계산
            score = 0
            if p.merchant_name == merchant_name:
                score = 100
            elif p.merchant_name in merchant_name:
                score = 80
            elif merchant_name in p.merchant_name:
                score = 60
            elif any(word in merchant_name for word in p.merchant_name.split()):
                score = 40

            if score > 0:
                suggestions.append({
                    "pattern_id": p.id,
                    "merchant_name": p.merchant_name,
                    "usage_description": p.usage_description,
                    "score": score,
                    "is_card_specific": p.card_id == card_id,
                })

        # 점수순 정렬
        suggestions.sort(key=lambda x: (-x["score"], -x.get("is_card_specific", False)))
        return suggestions[:5]  # 상위 5개

File: frontend/src/lib/matching.ts
import { supabase } from './supabase';

interface MatchResult {
  usageDescription: string;
  patternId: number;
}

/**
 * 패턴 매칭 로직 (3단계 우선순위)
 * 1. 카드 전용 정확 매칭 (card_id가 있는 exact 패턴)
 * 2. 공통 정확 매칭 (card_id가 NULL인 exact 패턴)
 * 3. 포함 매칭 (contains 패턴)
 */
export async function findMatch(
  merchantName: string,
  cardId: number
): Promise<MatchResult | null> {
  // 1. 카드 전용 정확 매칭
  const { data: cardExact } = await supabase
    .from('patterns')
    .select('*')
    .eq('card_id', cardId)
    .eq('match_type', 'exact')
    .eq('merchant_name', merchantName)
    .order('priority', { ascending: false })
    .limit(1);

  if (cardExact && cardExact.length > 0) {
    await incrementUseCount(cardExact[0].id);
    return {
      usageDescription: cardExact[0].usage_description,
      patternId: cardExact[0].id,
    };
  }

  // 2. 공통 정확 매칭
  const { data: commonExact } = await supabase
    .from('patterns')
    .select('*')
    .is('card_id', null)
    .eq('match_type', 'exact')
    .eq('merchant_name', merchantName)
    .order('priority', { ascending: false })
    .limit(1);

  if (commonExact && commonExact.length > 0) {
    await incrementUseCount(commonExact[0].id);
    return {
      usageDescription: commonExact[0].usage_description,
      patternId: commonExact[0].id,
    };
  }

  // 3. 포함 매칭
  const { data: containsPatterns } = await supabase
    .from('patterns')
    .select('*')
    .eq('match_type', 'contains')
    .order('priority', { ascending: false });

  if (containsPatterns) {
    for (const pattern of containsPatterns) {
      if (merchantName.includes(pattern.merchant_name)) {
        // 카드 전용 또는 공통 패턴만 적용
        if (pattern.card_id === null || pattern.card_id === cardId) {
          await incrementUseCount(pattern.id);
          return {
            usageDescription: pattern.usage_description,
            patternId: pattern.id,
          };
        }
      }
    }
  }

  return null;
}

/**
 * 패턴 사용 횟수 증가
 */
async function incrementUseCount(patternId: number): Promise<void> {
  try {
    // RPC로 use_count 증가 시도
    const { error } = await supabase.rpc('increment_pattern_use_count', { pattern_id: patternId });
    if (error) {
      // RPC가 없으면 직접 업데이트
      await supabase
        .from('patterns')
        .update({ use_count: 1 })
        .eq('id', patternId);
    }
  } catch {
    // 실패 시 무시 (비필수 기능)
  }
}

/**
 * 거래 일괄 매칭
 */
export async function batchMatch(
  transactions: Array<{ id: number; merchantName: string; cardId: number }>
): Promise<{ matched: number; failed: number }> {
  let matched = 0;
  let failed = 0;

  for (const tx of transactions) {
    const result = await findMatch(tx.merchantName, tx.cardId);
    if (result) {
      await supabase
        .from('transactions')
        .update({
          usage_description: result.usageDescription,
          matched_pattern_id: result.patternId,
          match_status: 'auto',
        })
        .eq('id', tx.id);
      matched++;
    } else {
      failed++;
    }
  }

  return { matched, failed };
}

/**
 * 패턴 제안 (유사 패턴 검색)
 */
export async function suggestPatterns(
  merchantName: string,
  cardId?: number
): Promise<Array<{
  patternId: number;
  merchantName: string;
  usageDescription: string;
  score: number;
}>> {
  let query = supabase.from('patterns').select('*');

  if (cardId) {
    query = query.or(`card_id.eq.${cardId},card_id.is.null`);
  }

  const { data: patterns } = await query;
  if (!patterns) return [];

  const suggestions: Array<{
    patternId: number;
    merchantName: string;
    usageDescription: string;
    score: number;
  }> = [];

  for (const pattern of patterns) {
    let score = 0;

    if (pattern.merchant_name === merchantName) {
      score = 100;
    } else if (pattern.merchant_name.includes(merchantName)) {
      score = 80;
    } else if (merchantName.includes(pattern.merchant_name)) {
      score = 60;
    } else {
      // 단어 단위 매칭
      const words = pattern.merchant_name.split(/\s+/);
      for (const word of words) {
        if (word.length > 2 && merchantName.includes(word)) {
          score = Math.max(score, 40);
        }
      }
    }

    if (score > 0) {
      suggestions.push({
        patternId: pattern.id,
        merchantName: pattern.merchant_name,
        usageDescription: pattern.usage_description,
        score,
      });
    }
  }

  // 점수순 정렬 후 상위 5개
  return suggestions
    .sort((a, b) => b.score - a.score)
    .slice(0, 5);
}

/**
 * 수동 매칭 시 패턴 저장
 */
export async function savePatternFromManualMatch(
  merchantName: string,
  usageDescription: string,
  cardId?: number
): Promise<void> {
  // 기존 패턴 확인
  let query = supabase
    .from('patterns')
    .select('id')
    .eq('merchant_name', merchantName)
    .eq('match_type', 'exact');

  if (cardId) {
    query = query.eq('card_id', cardId);
  } else {
    query = query.is('card_id', null);
  }

  const { data: existing } = await query.limit(1);

  if (existing && existing.length > 0) {
    // 기존 패턴 업데이트
    await supabase
      .from('patterns')
      .update({ usage_description: usageDescription })
      .eq('id', existing[0].id);
  } else {
    // 새 패턴 생성
    await supabase.from('patterns').insert({
      merchant_name: merchantName,
      usage_description: usageDescription,
      card_id: cardId || null,
      match_type: 'exact',
      priority: cardId ? 10 : 0,
      use_count: 1,
    });
  }
}
